### ISSUE TITLE
Can I use NCRF for punctuation restoration?

### ISSUE THREAD

**Issue by nguyenvanhieuvn** [jiesutd/NCRFpp#145]
I have try to train punctuation restoration using NCRFpp, my data has three label type:

- C: comma (,)
- P: period (.)
- and O

For example:
```
các     O
cơ_quan O
thông_tấn       O
báo_chí C
các     O
cơ_quan_chức_năng       O
có      O
thẩm_quyền      C
chị     O
tuyết   O
cho_hay P
sở_dĩ   O
các     O
gia_đình        O

phải    O
đưa     O
ra      O
yêu_cầu O
trên    O
khi     O
sự_việc O
đã      O
diễn    O
ra      O
từ      O
tháng   O
<NUM>   O
là      O
bởi     C
nguyên_nhân     O
của     O
tai_biến        O
được    O

```

But when training, i always get precision, recall and f1 = -1. And after 20 epochs training, decode result almost give me label `O`.

Here is my training logs:

```
Seed num: 42
MODEL: train
Load pretrained word embedding, norm: False, dir: /data/hieunv11/resources/cc.vi.300.vec
Embedding:
     pretrain word:1, prefect match:0, case_match:0, oov:175965, oov%:0.9999943170839821
Training model...
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
DATA SUMMARY START:
 I/O:
     Start   Sequence   Laebling   task...
     Tag          scheme: NoSeg
     Split         token:  ||| 
     MAX SENTENCE LENGTH: 250
     MAX   WORD   LENGTH: -1
     Number   normalized: True
     Word  alphabet size: 175966
     Char  alphabet size: 313
     Label alphabet size: 5
     Word embedding  dir: /data/hieunv11/resources/cc.vi.300.vec
     Char embedding  dir: None
     Word embedding size: 1
     Char embedding size: 50
     Norm   word     emb: False
     Norm   char     emb: False
     Train  file directory: data/punc_500k/train.txt
     Dev    file directory: data/punc_500k/dev.txt
     Test   file directory: data/punc_500k/test.txt
     Raw    file directory: None
     Dset   file directory: None
     Model  file directory: data/punc_500k/Model_500K_l60_e5_b128_lr0.015
     Loadmodel   directory: None
     Decode file directory: None
     Train instance number: 500000
     Dev   instance number: 100000
     Test  instance number: 100000
     Raw   instance number: 0
     FEATURE num: 0
 ++++++++++++++++++++++++++++++++++++++++
 Model Network:
     Model        use_crf: True
     Model word extractor: LSTM
     Model       use_char: True
     Model char extractor: LSTM
     Model char_hidden_dim: 100
 ++++++++++++++++++++++++++++++++++++++++
 Training:
     Optimizer: SGD
     Iteration: 20
     BatchSize: 128
     Average  batch   loss: True
 ++++++++++++++++++++++++++++++++++++++++
 Hyperparameters:
     Hyper              lr: 0.015
     Hyper        lr_decay: 0.05
     Hyper         HP_clip: None
     Hyper        momentum: 0.0
     Hyper              l2: 1e-08
     Hyper      hidden_dim: 200
     Hyper         dropout: 0.5
     Hyper      lstm_layer: 2
     Hyper          bilstm: True
     Hyper             GPU: True
DATA SUMMARY END.
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
build sequence labeling network...
use_char:  True
char feature extractor:  LSTM
word feature extractor:  LSTM
use crf:  True
build word sequence feature extractor: LSTM...
build word representation...
build char sequence feature extractor: LSTM ...
build CRF...
Epoch: 0/20
 Learning rate is set as: 0.015
Shuffle: first input word list: [35, 2468, 265, 2081, 260, 50, 733, 32, 1279, 756, 130, 715, 767, 484, 70808, 103, 235, 90, 1731, 1289, 50, 2670, 127, 2193, 1014, 13, 2081, 260, 95, 6, 2081, 260, 1463, 1536, 1048, 50, 8, 845, 1468, 683, 1371, 13, 2081, 260, 941, 486, 2081, 1250, 50, 1233]
     Instance: 16000; Time: 63.64s; loss: 3041.6690; acc: 758302.0/857935.0=0.8839
     Instance: 32000; Time: 63.64s; loss: 2941.0250; acc: 1516809.0/1715825.0=0.8840
     Instance: 48000; Time: 64.02s; loss: 2932.0961; acc: 2275832.0/2573962.0=0.8842
     Instance: 64000; Time: 65.92s; loss: 2923.5568; acc: 3035532.0/3432455.0=0.8844
     Instance: 80000; Time: 65.19s; loss: 2923.1116; acc: 3795061.0/4290926.0=0.8844
     Instance: 96000; Time: 66.57s; loss: 2923.5491; acc: 4554563.0/5149377.0=0.8845
     Instance: 112000; Time: 64.41s; loss: 2921.3159; acc: 5314156.0/6007805.0=0.8845
     Instance: 128000; Time: 63.10s; loss: 2919.0304; acc: 6073784.0/6866245.0=0.8846
     Instance: 144000; Time: 64.49s; loss: 2911.2097; acc: 6833904.0/7724904.0=0.8847
     Instance: 160000; Time: 63.42s; loss: 2917.5910; acc: 7593191.0/8583167.0=0.8847
     Instance: 176000; Time: 63.86s; loss: 2926.6022; acc: 8351881.0/9441179.0=0.8846
     Instance: 192000; Time: 64.23s; loss: 2913.9716; acc: 9111608.0/10299689.0=0.8846
     Instance: 208000; Time: 65.33s; loss: 2919.4587; acc: 9870589.0/11157849.0=0.8846
     Instance: 224000; Time: 64.01s; loss: 2923.2678; acc: 10629361.0/12015899.0=0.8846
     Instance: 240000; Time: 65.41s; loss: 2915.2301; acc: 11388427.0/12874062.0=0.8846
     Instance: 256000; Time: 64.04s; loss: 2901.1030; acc: 12148550.0/13732761.0=0.8846
     Instance: 272000; Time: 64.94s; loss: 2909.8870; acc: 12907175.0/14590728.0=0.8846
     Instance: 288000; Time: 64.97s; loss: 2893.9468; acc: 13666073.0/15448838.0=0.8846
     Instance: 304000; Time: 64.88s; loss: 2872.1600; acc: 14425199.0/16307047.0=0.8846
     Instance: 320000; Time: 63.61s; loss: 2856.7969; acc: 15184475.0/17165352.0=0.8846
     Instance: 336000; Time: 65.01s; loss: 2832.8147; acc: 15944170.0/18023834.0=0.8846
     Instance: 352000; Time: 62.62s; loss: 2819.3860; acc: 16703870.0/18882321.0=0.8846
     Instance: 368000; Time: 64.46s; loss: 2820.7084; acc: 17462509.0/19740233.0=0.8846
     Instance: 384000; Time: 63.22s; loss: 2799.3792; acc: 18221942.0/20598594.0=0.8846
     Instance: 400000; Time: 64.16s; loss: 2788.0835; acc: 18981422.0/21456976.0=0.8846
     Instance: 416000; Time: 64.24s; loss: 2784.1699; acc: 19740093.0/22314930.0=0.8846
     Instance: 432000; Time: 63.83s; loss: 2776.3037; acc: 20498793.0/23172858.0=0.8846
     Instance: 448000; Time: 65.09s; loss: 2767.3186; acc: 21258352.0/24031239.0=0.8846
     Instance: 464000; Time: 64.82s; loss: 2750.6462; acc: 22018088.0/24889676.0=0.8846
     Instance: 480000; Time: 63.09s; loss: 2747.9514; acc: 22778373.0/25748361.0=0.8847
     Instance: 496000; Time: 63.41s; loss: 2748.6209; acc: 23537776.0/26606494.0=0.8847
     Instance: 500000; Time: 16.32s; loss: 703.8519; acc: 23727737.0/26821058.0=0.8847
     Instance: 500000; Time: 0.07s; loss: 0.0000; acc: 23727737.0/26821058.0=0.8847
Epoch: 0 training finished. Time: 2010.01s, speed: 248.75st/s,  total loss: 89725.81311035156
totalloss: 89725.81311035156
Right token =  4759157  All token =  5369601  acc =  0.8863148304687816
Dev: time: 317.45s, speed: 319.60st/s; acc: 0.8863, p: -1.0000, r: -1.0000, f: -1.0000
Exceed previous best f score: -10
Save current best model in file: data/punc_500k/Model_500K_l60_e5_b128_lr0.015.0.model
Right token =  4759991  All token =  5369818  acc =  0.8864343260795804
Test: time: 310.25s, speed: 327.13st/s; acc: 0.8864, p: -1.0000, r: -1.0000, f: -1.0000
Epoch: 10/20
 Learning rate is set as: 0.01
Shuffle: first input word list: [710, 14314, 21, 125, 3992, 401, 3352, 367, 1014, 8, 17, 1507, 26457, 2042, 8, 39, 710, 14314, 1273, 501, 820, 3783, 63, 154, 1166, 683, 4376, 133, 619, 710, 108, 87, 1991, 278, 104, 1014, 8, 17, 1507, 1003, 235, 55, 21, 228, 5057, 683, 710, 439, 661, 31182]
     Instance: 16000; Time: 65.25s; loss: 2130.6108; acc: 770361.0/858528.0=0.8973
     Instance: 32000; Time: 66.32s; loss: 2140.8114; acc: 1539868.0/1716623.0=0.8970
     Instance: 48000; Time: 64.30s; loss: 2128.9162; acc: 2309662.0/2574768.0=0.8970
     Instance: 64000; Time: 64.31s; loss: 2130.1395; acc: 3079751.0/3433027.0=0.8971
     Instance: 80000; Time: 63.63s; loss: 2130.4807; acc: 3849646.0/4291127.0=0.8971
     Instance: 96000; Time: 64.24s; loss: 2127.5371; acc: 4619325.0/5149126.0=0.8971
     Instance: 112000; Time: 65.07s; loss: 2125.4153; acc: 5389498.0/6007339.0=0.8972
     Instance: 128000; Time: 66.51s; loss: 2128.4081; acc: 6159680.0/6865680.0=0.8972
     Instance: 144000; Time: 64.02s; loss: 2121.3087; acc: 6930260.0/7724348.0=0.8972
     Instance: 160000; Time: 64.50s; loss: 2118.4020; acc: 7700703.0/8582883.0=0.8972
     Instance: 176000; Time: 64.92s; loss: 2120.1167; acc: 8471598.0/9441538.0=0.8973
     Instance: 192000; Time: 65.60s; loss: 2123.6754; acc: 9241800.0/10299774.0=0.8973
     Instance: 208000; Time: 64.70s; loss: 2118.5656; acc: 10011908.0/11157915.0=0.8973
     Instance: 224000; Time: 64.68s; loss: 2121.3178; acc: 10782028.0/12016264.0=0.8973
     Instance: 240000; Time: 65.54s; loss: 2130.8972; acc: 11551345.0/12874144.0=0.8973
     Instance: 256000; Time: 62.71s; loss: 2110.7265; acc: 12322231.0/13732674.0=0.8973
     Instance: 272000; Time: 65.18s; loss: 2115.0816; acc: 13092882.0/14591225.0=0.8973
     Instance: 288000; Time: 63.90s; loss: 2112.8263; acc: 13863313.0/15449455.0=0.8973
     Instance: 304000; Time: 64.67s; loss: 2122.5491; acc: 14633518.0/16307724.0=0.8973
     Instance: 320000; Time: 65.50s; loss: 2112.3600; acc: 15404386.0/17166151.0=0.8974
     Instance: 336000; Time: 65.21s; loss: 2114.0927; acc: 16174526.0/18024258.0=0.8974
     Instance: 352000; Time: 64.30s; loss: 2115.0721; acc: 16945081.0/18882485.0=0.8974
     Instance: 368000; Time: 65.43s; loss: 2112.4929; acc: 17715874.0/19740954.0=0.8974
     Instance: 384000; Time: 65.30s; loss: 2112.1071; acc: 18485625.0/20598837.0=0.8974
     Instance: 400000; Time: 64.66s; loss: 2102.5770; acc: 19256604.0/21457316.0=0.8974
     Instance: 416000; Time: 65.99s; loss: 2105.8262; acc: 20027315.0/22315530.0=0.8975
     Instance: 432000; Time: 65.00s; loss: 2111.2730; acc: 20797778.0/23173631.0=0.8975
     Instance: 448000; Time: 65.43s; loss: 2112.6652; acc: 21568509.0/24031935.0=0.8975
     Instance: 464000; Time: 65.37s; loss: 2111.9879; acc: 22338789.0/24890041.0=0.8975
     Instance: 480000; Time: 64.01s; loss: 2108.8273; acc: 23109242.0/25748150.0=0.8975
     Instance: 496000; Time: 64.67s; loss: 2099.3821; acc: 23880093.0/26606636.0=0.8975
     Instance: 500000; Time: 16.28s; loss: 542.0894; acc: 24072398.0/26821058.0=0.8975
     Instance: 500000; Time: 0.12s; loss: 0.0000; acc: 24072398.0/26821058.0=0.8975
Epoch: 10 training finished. Time: 2027.36s, speed: 246.63st/s,  total loss: 66218.53881835938
totalloss: 66218.53881835938
Right token =  4837242  All token =  5369601  acc =  0.9008568793100269
Dev: time: 319.31s, speed: 317.78st/s; acc: 0.9009, p: -1.0000, r: -1.0000, f: -1.0000
Right token =  4840480  All token =  5369818  acc =  0.9014234746875965
Test: time: 325.79s, speed: 311.42st/s; acc: 0.9014, p: -1.0000, r: -1.0000, f: -1.0000
Epoch: 19/20
 Learning rate is set as: 0.007692307692307691
Shuffle: first input word list: [710, 744, 373, 444, 1083, 293, 86, 21, 17, 18, 13, 339, 1321, 130, 270, 630, 62, 249, 96, 176, 619, 153, 176, 152, 710, 104, 1228, 710, 7803, 85, 293, 309, 90, 1141, 4159, 17, 18, 339, 1321, 62, 293, 86, 21, 1689, 1083, 4159, 3499, 104, 367, 190, 270, 245]
     Instance: 16000; Time: 66.62s; loss: 1944.9768; acc: 775043.0/857887.0=0.9034
     Instance: 32000; Time: 66.42s; loss: 1947.0708; acc: 1549484.0/1715435.0=0.9033
     Instance: 48000; Time: 67.35s; loss: 1932.2689; acc: 2325524.0/2573732.0=0.9036
     Instance: 64000; Time: 66.90s; loss: 1936.6950; acc: 3101176.0/3432033.0=0.9036
     Instance: 80000; Time: 66.07s; loss: 1939.3614; acc: 3876885.0/4290350.0=0.9036
     Instance: 96000; Time: 66.33s; loss: 1934.6481; acc: 4652677.0/5148773.0=0.9036
     Instance: 112000; Time: 65.91s; loss: 1927.1535; acc: 5429219.0/6007348.0=0.9038
     Instance: 128000; Time: 66.63s; loss: 1938.4611; acc: 6205061.0/6865701.0=0.9038
     Instance: 144000; Time: 65.93s; loss: 1934.6078; acc: 6980898.0/7724053.0=0.9038
     Instance: 160000; Time: 66.58s; loss: 1939.7623; acc: 7756183.0/8582086.0=0.9038
     Instance: 176000; Time: 66.42s; loss: 1925.4272; acc: 8532296.0/9440309.0=0.9038
     Instance: 192000; Time: 66.29s; loss: 1922.9198; acc: 9308351.0/10298417.0=0.9039
     Instance: 208000; Time: 66.62s; loss: 1932.7795; acc: 10084175.0/11156799.0=0.9039
     Instance: 224000; Time: 66.69s; loss: 1932.2163; acc: 10860466.0/12015373.0=0.9039
     Instance: 240000; Time: 67.85s; loss: 1929.5265; acc: 11636274.0/12873610.0=0.9039
     Instance: 256000; Time: 65.75s; loss: 1920.0203; acc: 12413098.0/13732363.0=0.9039
     Instance: 272000; Time: 65.83s; loss: 1927.3521; acc: 13189516.0/14591026.0=0.9039
     Instance: 288000; Time: 65.30s; loss: 1929.7130; acc: 13965128.0/15449035.0=0.9039
     Instance: 304000; Time: 63.53s; loss: 1932.1134; acc: 14740425.0/16306959.0=0.9039
     Instance: 320000; Time: 64.63s; loss: 1929.8634; acc: 15516462.0/17165245.0=0.9039
     Instance: 336000; Time: 64.98s; loss: 1932.6875; acc: 16292036.0/18023466.0=0.9039
     Instance: 352000; Time: 63.50s; loss: 1920.0932; acc: 17069098.0/18882307.0=0.9040
     Instance: 368000; Time: 63.25s; loss: 1917.1590; acc: 17845565.0/19740687.0=0.9040
     Instance: 384000; Time: 64.57s; loss: 1922.8176; acc: 18621545.0/20598849.0=0.9040
     Instance: 400000; Time: 66.21s; loss: 1929.0937; acc: 19397537.0/21457130.0=0.9040
     Instance: 416000; Time: 63.81s; loss: 1931.6431; acc: 20173232.0/22315290.0=0.9040
     Instance: 432000; Time: 65.64s; loss: 1921.5123; acc: 20949871.0/23173981.0=0.9040
     Instance: 448000; Time: 64.62s; loss: 1926.5331; acc: 21725649.0/24031980.0=0.9040
     Instance: 464000; Time: 64.11s; loss: 1921.4868; acc: 22501961.0/24890363.0=0.9040
     Instance: 480000; Time: 63.54s; loss: 1927.1855; acc: 23277934.0/25748570.0=0.9040
     Instance: 496000; Time: 64.07s; loss: 1922.5005; acc: 24053500.0/26606436.0=0.9040
     Instance: 500000; Time: 16.74s; loss: 488.5893; acc: 24247726.0/26821058.0=0.9041
     Instance: 500000; Time: 0.07s; loss: 0.0000; acc: 24247726.0/26821058.0=0.9041
Epoch: 19 training finished. Time: 2048.75s, speed: 244.05st/s,  total loss: 60318.23893737793
totalloss: 60318.23893737793
Right token =  4873121  All token =  5369601  acc =  0.9075387538105718
Dev: time: 316.09s, speed: 321.13st/s; acc: 0.9075, p: -1.0000, r: -1.0000, f: -1.0000
Right token =  4876656  All token =  5369818  acc =  0.9081603883036632
Test: time: 346.18s, speed: 292.97st/s; acc: 0.9082, p: -1.0000, r: -1.0000, f: -1.0000

```

What am i wrong?

**Comment by jiesutd** [comment]
1. This is not a Segmentation or NER problem, so the P/R/F metric is not proper here. You need to set ‘seg=False’ in the configuration file.

2. Your pretrained embeddings didn’t load correctly, the matched word is zero. Double check the format of pretrained embeddings.
